In [1]:
import numpy as np
import pandas as pd

### Q1. Pandas version

In [2]:
pd.__version__

'2.3.2'

### Q2. Records count 

In [3]:
#!wget https://raw.githubusercontent.com/alexeygrigorev/datasets/master/car_fuel_efficiency.csv

In [4]:
df = pd.read_csv('data/car_fuel_efficiency.csv')
print(df.shape)
df.head()

(9704, 11)


engine_displacement  num_cylinders  horsepower  vehicle_weight  \
0                  170            3.0       159.0     3413.433759   
1                  130            5.0        97.0     3149.664934   
2                  170            NaN        78.0     3079.038997   
3                  220            4.0         NaN     2542.392402   
4                  210            1.0       140.0     3460.870990   

   acceleration  model_year  origin fuel_type         drivetrain  num_doors  \
0          17.7        2003  Europe  Gasoline    All-wheel drive        0.0   
1          17.8        2007     USA  Gasoline  Front-wheel drive        0.0   
2          15.1        2018  Europe  Gasoline  Front-wheel drive        0.0   
3          20.2        2009     USA    Diesel    All-wheel drive        2.0   
4          14.4        2009  Europe  Gasoline    All-wheel drive        2.0   

   fuel_efficiency_mpg  
0            13.231729  
1            13.688217  
2            14.246341  
3            16.912736  
4            12.488369

There are 9704 records in the dataset

### Q3. Fuel types

In [5]:
df['fuel_type'].value_counts()

fuel_type
Gasoline    4898
Diesel      4806
Name: count, dtype: int64

In [6]:
df['fuel_type'].nunique()

2

There are 2 fuel types presented in the dataset

### Q4. Missing values

In [7]:
df.isna().any().sum()

np.int64(4)

There are 4 columns in the dataset with missing values

### Q5. Max fuel efficiency

What's the maximum fuel efficiency of cars from Asia?

In [8]:
df.query("origin == 'Asia'")['fuel_efficiency_mpg'].max()

23.759122836520497

### Q6. Median value of horsepower

In [9]:
# Find the median value of the horsepower column in the dataset.
df['horsepower'].median()

149.0

In [10]:
# Next, calculate the most frequent value of the same horsepower column.
horsepower_most_frequent = df['horsepower'].mode()[0]
horsepower_most_frequent

np.float64(152.0)

In [11]:
# Use the fillna method to fill the missing values in the horsepower column 
# with the most frequent value from the previous step.
df['horsepower'] = df['horsepower'].fillna(horsepower_most_frequent)

In [12]:
# Now, calculate the median value of horsepower once again.
df['horsepower'].median()

152.0

Median value of horsepower is increased

### Q7. Sum of weights

In [13]:
# Select all the cars from Asia
df_asia = df.query("origin == 'Asia'")

In [15]:
# Select only columns vehicle_weight and model_year
df_asia = df_asia[['vehicle_weight','model_year']]

In [16]:
# Select the first 7 values
df_asia = df_asia.head(7)
df_asia

vehicle_weight  model_year
8      2714.219310        2016
12     2783.868974        2010
14     3582.687368        2007
20     2231.808142        2011
21     2659.431451        2016
34     2844.227534        2014
38     3761.994038        2019

In [18]:
# Get the underlying NumPy array. Let's call it X.
X = df_asia.values

In [21]:
# Compute matrix-matrix multiplication between the transpose of X and X. 
# To get the transpose, use X.T. Let's call the result XTX
XTX = X.T@X
XTX

array([[62248334.33150762, 41431216.5073268 ],
       [41431216.5073268 , 28373339.        ]])

In [23]:
# Invert XTX
XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 5.71497081e-07, -8.34509443e-07],
       [-8.34509443e-07,  1.25380877e-06]])

In [24]:
# Create an array y with values [1100, 1300, 800, 900, 1000, 1100, 1200]
y = [1100, 1300, 800, 900, 1000, 1100, 1200]

In [30]:
# Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. 
# Call the result w
temp = XTX_inv@X.T
w = temp@y
w

array([0.01386421, 0.5049067 ])

In [31]:
# What's the sum of all the elements of the result?
w.sum()

np.float64(0.5187709081074023)

0.51 is the sum of all the elements of the result